[Jaryl Chew] iP

techjay-c opened the pull request:
### Novo Tracking App

> "Track everything, all at once."

Novo manages your tasks all in one place, be in events, to do's or deadlines. It's

- Easy to use
- Simple to use
- ~~Minimalistic~~ Super Minimalistic

All you have to do is,

1. Download it [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html)
2. Run it
3. Use it

IT IS **FREE**!!

Features:

- [x] Task Marking
- [x] Searching for tasks
- [x] Managing Events
- [x] Managing Deadlines
- [x] Managing To-Do's
- [ ] Graphical User Interface (coming soon)

You can practice Java too: 
`public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}`

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
Duke Level 0

Commit:
Duke Level 1

Commit:
Duke Level 2

Commit:
Duke Level 3

Commit:
Duke Level 4

Commit:
Duke A-TextUiTesting

Commit:
Duke Level-5

Commit:
Duke Level-6

Commit:
Duke A-Enums

Commit:
Duke Level-7

Commit:
Duke Level-8

Commit:
Duke A-MoreOOP

Commit:
Duke A-Packages

Commit:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2324S1/ip

* 'add-gradle-support' of https://github.com/nus-cs2103-AY2324S1/ip:
  Bump gradle and lib version
  Add Gradle support

Commit:
A-Gradle

Commit:
Add JUnit tests for TaskList and Todo

Commit:
Created JAR file

Commit:
Add JavaDoc comments to all classes

Commit:
Update classes to comply with coding standard

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Add Search feature to allow users to search for specific task

Commit:
Add remaining files for Search feature
- This commit includes the files that were missed in the previous "Add Search feature" commit.

Commit:
Merge branch 'branch-Level-9'

Commit:
Update files to adhere to coding style requirements

Commit:
Implemented JavaFX as a separate project

marioalvaro commented on data/duke.txt:
Next commit can try to remove this file.

marioalvaro commented on src/main/java/duke/Storage.java:
```diff
@@ -54,18 +53,18 @@ public static void saveTasks(ArrayList<Task> tasks) {
 
                 if (taskType.equals("Event")) {
```
This method is using arrowstyle method, try to avoid it

marioalvaro commented on src/main/java/duke/Storage.java:
```diff
@@ -54,18 +53,18 @@ public static void saveTasks(ArrayList<Task> tasks) {
+                    LocalDateTime fromDateTime = event.fromDateAndTime();
+                    LocalDateTime toDateTime = event.toDateAndTime();
                     String formattedFromDateTime = fromDateTime.format(DateTimeFormatter.ofPattern("MMMM d yyyy ha"));
                     String formattedToDateTime = toDateTime.format(DateTimeFormatter.ofPattern("MMMM d yyyy ha"));
```
Preferably add space between the Event, LocalDateTime, and String variable to create ease of reading

marioalvaro commented on src/main/java/duke/Storage.java:
```diff
@@ -54,18 +53,18 @@ public static void saveTasks(ArrayList<Task> tasks) {
-                    writer.write("E " + "| " + statusIcon + " | " + event.description +
-                            " | " + formattedFromDateTime + " - " + formattedToDateTime + "\n");
+                    writer.write("E " + "| " + statusIcon + " | " + event.description
+                        + " | " + formattedFromDateTime + " - " + formattedToDateTime + "\n");
```
Indentation should be 8 spaces.

marioalvaro commented on src/main/java/duke/Storage.java:
```diff
@@ -76,9 +75,11 @@ public static void saveTasks(ArrayList<Task> tasks) {
     // Solution below adapted and inspired by https://www.guru99.com/buffered-reader-in-java.html
     // Solution below adapted and inspired by https://chat.openai.com/share/4f6c03e6-99d5-47c0-8887-1762a36b15fb
+
     /**
```
This method also uses arrowstyle method, try to avoid it.

marioalvaro commented on src/main/java/duke/Storage.java:
```diff
@@ -94,7 +95,8 @@ static ArrayList<Task> loadTasks() throws IOException {
                 if (file.getParentFile() != null) {
                     // Solution below adapted from
-                    // https://www.oreilly.com/library/view/java-cookbook/0596001703/ch10s10.html#:~:text=Of%20the%20two%20methods%20used,%2Fian%2Fbin%22).
+                    // https://www.oreilly.com/library/view/java-cookbook/0596001703/ch10s10.html#:~:
```
Nice for the information regarding getting the idea!

marioalvaro commented on src/main/java/duke/Ui.java:
```diff
@@ -171,14 +171,17 @@ public void printUnmarkException() {
+     * Displays an error message for a search command in the wrong format.
      */
     public void printSearchException() {
         System.out.println("____________________________________________________________");
```
why not use separators constant for the line?

marioalvaro commented on src/main/java/duke/Ui.java:
```diff
@@ -9,8 +9,8 @@
  */
 public class Ui {
     private static final String separators = "____________________________________________________________";
```
Refer back to the Coding Standard for static final constant, all letters should be capitalized.

marioalvaro commented on src/main/java/duke/Duke.java:
```diff
@@ -37,7 +40,8 @@ public class Duke {
-    public static void main(String[] args) throws DukeException.NoSuchItemException, DukeException.ToDoException, IOException {
+    public static void main(String[] args) throws DukeException.NoSuchItemException,
+            DukeException.ToDoException, IOException {
         ArrayList<Task> tasks = new ArrayList<>();
```
Why not use TaskList?

marioalvaro reviewed (commented):
Overall nice implementation.
But there are several indentations that are incorrect.
Otherwise, can consider adding author in each public Class.
Noce Job

Commit:
Update Duke to implement GUI

Commit:
Merge branch 'branch-Level-10'

s0ngyang commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,79 @@
+        this.description = description;
+        this.isDone = false;
+        this.taskType = taskType;
+    }
+
+    /**
+     * Gets the status icon of the task.
+     *
+     * @return "X" if the task is done, " " (a space) if the task is not done.
+     */
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
+    }
+
+    /**
+     * Marks the task as done.
+     */
+    public void markAsDone() {
+        this.isDone = true;
+    }
+
+    /**
+     * Marks the task as not done.
+     */
+    public void markAsNotDone() {
+        this.isDone = false;
+    }
+
+    /**
+     * Gets the description of the task.
+     *
+     * This method returns the description of the task as a String.
+     *
+     * @return The description of the task as a String.
+     */
+    public String getDescription() {
+        return this.description;
+    }
+}
+
```
Perhaps Task.java could be organised under the task directory?

s0ngyang commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,182 @@
+
+        mainLayout.setPrefSize(400.0, 600.0);
+
+        scrollPane.setPrefSize(385, 535);
+        scrollPane.setHbarPolicy(ScrollPane.ScrollBarPolicy.NEVER);
+        scrollPane.setVbarPolicy(ScrollPane.ScrollBarPolicy.ALWAYS);
+
+        scrollPane.setVvalue(1.0);
+        scrollPane.setFitToWidth(true);
+
+        // You will need to import `javafx.scene.layout.Region` for this.
+        dialogContainer.setPrefHeight(Region.USE_COMPUTED_SIZE);
+
+        userInput.setPrefWidth(325.0);
+
+        sendButton.setPrefWidth(55.0);
+
+        AnchorPane.setTopAnchor(scrollPane, 1.0);
+
+        AnchorPane.setBottomAnchor(sendButton, 1.0);
+        AnchorPane.setRightAnchor(sendButton, 1.0);
+
+        AnchorPane.setLeftAnchor(userInput, 1.0);
+        AnchorPane.setBottomAnchor(userInput, 1.0);
+
+        //Step 3. Add functionality to handle user input.
+        sendButton.setOnMouseClicked((event) -> {
+            dialogContainer.getChildren().add(getDialogLabel(userInput.getText()));
+            userInput.clear();
+        });
+
+        userInput.setOnAction((event) -> {
+            dialogContainer.getChildren().add(getDialogLabel(userInput.getText()));
+            userInput.clear();
+        });
+
+        //Scroll down to the end every time dialogContainer's height changes.
+        dialogContainer.heightProperty().addListener((observable) -> scrollPane.setVvalue(1.0));
+    }
+     */
```
Nit: Commented code could be removed, though it may be WIP?

s0ngyang commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,124 @@
+
+            DateTimeValidator validatorE = new DateTimeValidator("yyyy/MM/dd HHmm");
+            boolean isDateValidE = validatorE.validateDate(fromDateTime) && validatorE.validateDate(toDateTime);
+
+            if (isDateValidE) {
+                LocalDateTime parsedFromDate = LocalDateTime.parse(fromDateTime,
+                    DateTimeFormatter.ofPattern("yyyy/MM/dd HHmm"));
+                LocalDateTime parsedToDate = LocalDateTime.parse(toDateTime,
+                    DateTimeFormatter.ofPattern("yyyy/MM/dd HHmm"));
+                return new AddCommand(AddCommand.TaskType.EVENT, theDescriptionE, parsedFromDate, parsedToDate);
+            }
+        case "deadline":
+            // Solution below adapted and inspired by https://chat.openai.com/share/b706b4df-ab30-4d0f-93eb-b85617616319
+
+            if (arguments.isEmpty()) {
+                throw new DukeException.DeadlineException();
+            }
+
+            String userCommandD = userInput.split(" ")[0];
+            String argsD = userInput.replaceFirst(userCommandD, "").trim();
+            String[] splitTheArgumentsD = argsD.split("/by", 2);
+
+            if (splitTheArgumentsD.length != 2) {
+                throw new DukeException.DeadlineFormatException();
+            }
+
+            String theDescriptionD = splitTheArgumentsD[0];
+            String[] theDateTimeD = splitTheArgumentsD[1].trim().split(" ", 2);
+
+            String date = theDateTimeD[0];
+            String time = theDateTimeD[1];
+
+            DateTimeValidator validator = new DateTimeValidator("dd/MM/yyyy HHmm");
+            boolean isDateValid = validator.validateDate(date + " " + time);
+
+            if (isDateValid) {
+                LocalDateTime parsedDateTime = LocalDateTime.parse(date + " " + time,
+                    DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm"));
+                return new AddCommand(AddCommand.TaskType.DEADLINE, theDescriptionD, parsedDateTime);
+            }
```
Could the more complex cases such as event and deadline be abstracted out?

s0ngyang commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,124 @@
+    public static Command parseCommand(String userInput) throws DukeException {
+        String[] split = userInput.split(" ");
+        String command = split[0];
+        String arguments = userInput.replaceFirst(command, "").trim();
```
I like the use of `replaceFirst`, personally did not think of using that.

s0ngyang reviewed (commented):
LGTM, I like that you cited all your sources, and the emoji in the exception messages are nice. Overall, this is a clean and organised codebase!

wj331 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,182 @@
+            Command command = parser.parseCommand(input);
+            return command.execute(taskList, ui, storage);
+        } catch (DukeException.ToDoException e) {
+            return ui.printToDoException();
```
Too many duplicate code here.

Consider doing something like
`catch (DukeException.ToDoException e) {
      return ui.printException(e.getMessage())
 }`
 and store the message directly in your exception?

wj331 commented on src/main/java/duke/command/SearchCommand.java:
```diff
@@ -0,0 +1,50 @@
+     */
+    @Override
+    public String execute(TaskList taskList, Ui ui, Storage storage) {
+        ArrayList<Task> matchedTasks = new ArrayList<>();
```
As mentioned by someone previously, can consider using TaskList as well

wj331 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,164 @@
+
+        try {
+            File file = new File(filePath);
+            if (!file.exists()) {
```
seems like an arrowhead forming, maybe try to restructure your code?

wj331 reviewed (commented):
LGTM, generally readable and works. Some issues on coding style and quality to work on.

Commit:
Add Assertions

techjay-c committed:
Merge pull request #1 from techjay-c/branch-A-Assertions

Add Assertions

Commit:
Improve Code Quality

There are several sections in the project that can be improved further to achieve better code quality and readability.

Making these changes improves the code quality.

Changed the naming of certain static variables and added separate Parser classes for Event and Deadline.

techjay-c committed:
Merge pull request #2 from techjay-c/branch-A-CodeQuality

Improve Code Quality

Commit:
Add gradle.yml

Commit:
Enhance Search Feature

Provide more flexibility for users when searching tasks.

Enhanced search functionality for users.

Users can now search for tasks using keywords and specify which type of task to filter. If the keyword matches an item and the specified task type, the task will be returned.

techjay-c committed:
Merge pull request #4 from techjay-c/branch-C-BetterSearch

Enhance Search Feature

Commit:
Update files to comply with checkstyle

Commit:
Update Parser file to comply with checkstyle

techjay-c committed:
Merge pull request #5 from techjay-c/branch-C-BetterSearch

Update Parser file to comply with checkstyle

Commit:
Update Parser and Storage classes.

techjay-c committed:
Update README.md

techjay-c committed:
Update README.md

techjay-c committed:
Update README.md

techjay-c committed:
Update README.md

techjay-c committed:
Update README.md

techjay-c committed:
Update README.md

techjay-c committed:
Update README.md

techjay-c committed:
Update README.md

Commit:
Update J Unit Tests. Update Deadline and Event Parser classes.

Commit:
Update DeadlineTest and EventTest files

Commit:
Remove unused imports in Deadline Test to comply with checkstyle

Commit:
Update unit tests

Commit:
Add additional methods for unit testing

Commit:
Add Ui image

Commit:
Fix bugs

Commit:
Update Parser class. Changed Search command to use find keyword instead of search

techjay-c committed:
Update README.md

Commit:
Update Ui.png file to reflect find command
